.github/pull_request_template.md

```diff
@@ -15,9 +21,7 @@ Make sure you've run and fixed any issues with these commands:
 - `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
 - `cargo test --workspace --features=extra` to check that all tests pass
 
-# User-Facing Changes
-
-If you're making changes that will affect the user experience of Nushell (ex: adding/removing a command, changing an input/output type, adding a new flag):
+# After Submitting
 
-- Get another regular contributor to review the PR before merging
```

Comment:
I think Dan would be a little sad if this gets removed. He was quite keen on having this as a requirement, even for core team members.

Reply:
I deleted this because I wrote a similar sentiment to the "Before Submitting" subsection where I put a green light from core team as a requirement. Maybe we could formulate it as "seek feedback from regular contributors and get a green light from the core team"?

Reply:
I think that the "Before Submitting" requirement is fairly different from the "user-facing changes need design review" requirement.

There is some overlap, but the design review can catch small details that might be missed in a high-level pre-PR discussion.

Reply:
Should we then put into "After Submitting": "Get another regular contributor to review the PR"?

Reply:
@rgwood Check out the newest change. Is that what you had in mind?

Reply:
Hmm. I think that asking every change to be reviewed is quite different from asking for review on user-facing changes.

Looping in @dandavison as I know this is an area he has opinions about

Reply:
Thanks, the opinion I have is specifically about user-facing changes, like @kubouch wrote in the PR description:

> Adding/removing a command, changing the meaning of a command, changing an input or output type, changing flags or positional arguments.

I'd like to suggest that anyone (even core team members) making changes like that has to get approval from a regular contributor or core team member.

So I'm not talking about requiring general code review. I'm talking about something more like

```
### User-facing changes

This PR changes the user-facing API of the command `take` in two ways:

1. `my-list | take 1` now returns a list; it does not unwrap automatically
2. `my-list | take` is now an error: you must supply an argument (previously it was the same as `take 1`.
```

Then, that text there (not the whole code), would need approval by >=1 core team member / other approved contributor.

The reason I'm suggesting this is that historically the project has been treating the collection of nushell commands in `nu-command` as a suitable area for less experienced contributors to start making contributions to nushell in. But actually it's the opposite: `nu-command` should be the most locked-down area, because virtually any change there is a breaking user-facing change, and the standard library of commands is going to play a huge role in how the public judges nushell. We simply can't get to 1.0 until we're extremely happy with the design of that collection and it's going to take a lot of work.

This will be particularly important once we prune down the builtins to the core, and move others to community-contributed plugins, as suggested in https://github.com/nushell/nushell.github.io/blob/55484dfb610026f1071210fa993bce6df7967c1e/contributor-book/philosophy_0_80.md#core-categories

Reply:
@dandavison I see what you mean and I agree. I reformulated the text in both the contributing guide and the PR template to reflect that. The reason I keep it more general than just "User-Facing Features" is that it applies to the internals as well. For example, if someone comes up with a major redesign of the plugin protocol that doesn't quite fit in, it's better to discuss that beforehand. I'd say same rules should apply here as well.

@rgwood Are you both on board with the current formulations?

Feel free to edit if you disagree with something.

Reply:
OK, sorry I think I didn't look at this PR carefully enough. I think we're talking about two different things. The change you're making in this PR is about getting pre-approval for an idea. But I'm talking about getting agreement from colleagues on all user-facing API change, _prior to merge_ (whether the author sought prior approval or not). 

In what I'm talking about, there'd would be a completely black-and-white division between implementation and user-facing API, and the agreement from colleagues would only be for the user-facing API change (they are not reviewing the implementation). So if the author redesigned the plugin protocol in a way that affects the user experience, that would be user-facing API. Similarly any change to additions/deletions of builtin commands, or changes to command semantics or type signatures / flags / options etc would be user-facing API.

What I'm suggesting is that everyone, even if they have permission to push to `main` on their own, needs at least one other core contributor to agree with user-facing API changes. The idea is that this will help us evolve toward a shared agreement about design philosophy, and help ensure that we stick to the shared view on design philosophy.

And -- sorry to repeat myself! -- but it's only user-facing API changes that are relevant to design philosophy, by definition: our design philosophy (e.g. https://github.com/nushell/nushell.github.io/pull/656) concerns the user's experience of using nushell, and any change to that is defined to be a user-facing API change, as opposed to implementation/internals.

What I'm saying doesn't have to be represented by a change to the PR template: after all, people without merge permissions can't merge without core team approval. So technically, it would only be a core-team protocol change. But it might help to codify it in the PR template.

Reply:
I see. It seems you're proposing the following workflow:
1. I have an idea 
2. I discuss it on Discord and get a green light
3. I implement it and list all user-facing features in the PR description
4. Other contributors / core team members comment and suggest changes
5. I implement the changes
6. Everybody agrees
7. Core team member merges the PR

I'd say the user-facing change should already be a part of the initial idea, but I think the additional post-implementation step would help keeping it on track. 

I agree, we do not need to put it to the PR template to keep it short (there is already a requirement for listing the UX changes). We could, however, put it to the contributing guide.